fix(gnovm): make LeftmostX return true root of chains

audrenbdb opened the pull request:
Related issue: https://github.com/gnolang/gno/issues/4778

## Problem

- LeftmostX stopped after the first selector/index step, returning an intermediate node (e.g., a.b), not the real base (e.g., a).
- This caused downstream logic (notably escape analysis) to misidentify the owning variable and mark heap usage incorrectly.

## Solution

- Change LeftmostX to iteratively unwrap chains of: SelectorExpr (x.y) > IndexExpr (x[i]) > SliceExpr (x[:])

It still stops at boundaries we should not cross: callExpr, TypeAssertExpr, StarExpr (deref), RefExpr (address-of), etc.

Update the doc comment to clarify behavior and stopping rules.

### Consideration

Added SliceExpr to the LeftmostX. Why adding SliceExpr seems okay to me:

SliceExpr is part of the same selector/index “chain” family (e.g., a[:].b), so unwrapping it is consistent with the original intent and fixes real cases without reducing safety.

audrenbdb committed:
fix(gnovm): make LeftmostX return true root of chains

Problem

- LeftmostX stopped after the first selector/index step, returning an intermediate node (e.g., a.b), not the real base (e.g., a).
- This caused downstream logic (notably escape analysis) to misidentify the owning variable and mark heap usage incorrectly.

Solution

- Change LeftmostX to iteratively unwrap chains of: SelectorExpr (x.y) > IndexExpr (x[i]) > SliceExpr (x[:])

It still stops at boundaries we should not cross: callExpr, TypeAssertExpr, StarExpr (deref), RefExpr (address-of), etc.

Update the doc comment to clarify behavior and stopping rules.

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [x] The pull request description provides enough details (checked by @thehowl)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: audrenbdb/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🟢 User notJoon already reviewed PR 4824 with state APPROVED
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

notJoon reviewed (changes requested):
I didn't find any counterexamples in the fix. I think the implementation seems no problem. However to strengthen test coverage, I suggest adding the following cases:

1. `v.(T).field` should stop at `v.(T)`
2. `m[k].field` should correctly unwraps to `m`
3. `ptr.Method().field` should stop at `ptr.Method()`

After adding these, it should be good to go.

audrenbdb committed:
feat: cover LeftmostX function with additional tests

notJoon reviewed (approved):
LGTM. I've checked the major tests, but it would be better to verify the full regression through CI

Villaquiranm reviewed (approved):
Looks good to me :)
clean changes and tests

audrenbdb commented:
> LGTM. I've checked the major tests, but it would be better to verify the full regression through CI

@notJoon Thanks, is there anything I can trigger from my side regarding the CI ? I don't think I can but I may be missing something.

notJoon commented:
> > LGTM. I've checked the major tests, but it would be better to verify the full regression through CI
> 
> @notJoon Thanks, is there anything I can trigger from my side regarding the CI ? I don't think I can but I may be missing something.

@audrenbdb Due to the branch policy, first-time contributors need final approval from a maintainer (i.e., tech-staff) for the CI to run. Therefore, I think you don't need to worry about it right now.

notJoon committed:
Merge branch 'master' into incorrect-early-return-leftmostx

notJoon commented:
This PR also seems to require final review. cc: @gnolang/tech-staff

thehowl committed:
Merge remote-tracking branch 'upstream/master' into incorrect-early-return-leftmostx

thehowl committed:
fix(gnovm): fix compile error in TestLeftmostX and add filetest

- Pass stackRecursiveTraceLimit to toExprTrace calls in TestLeftmostX
- Add ref_nested_sel.gno filetest: verifies that &a.b.c heap-escapes the
  root variable `a` (HeapDefine/HeapUse annotations), not just `a.b`

thehowl commented:
@ltzmaxwell This seems like a correctness bug with no current real impact, though still probably good to merge. Here's what claude said:

> 
> ## Result: the bug is real but not runtime-observable
> 
> I built both binaries (pre-fix and post-fix) and ran every scenario I could construct. All produce the same output. The bug cannot be surfaced with a filetest.
> 
> **Why:** GnoVM's interpreter never actually has dangling pointers. When `&o.i.v` is evaluated, `PointerValue.Base` is set to `inner_sv` (the Go `*StructValue` for `o.i`), which the Go GC keeps alive for as long as the pointer is reachable — with or without `o` being a `HeapItemValue`.
> 
> **But the preprocessed form does differ.** Without the fix:
> ```
> func f() { o<!VPBlock(1,0)> := ...;  gp = &(o<VPBlock(1,0)>.i.v) }
> ```
> With the fix:
> ```
> func f() { o<!~VPBlock(1,0)> := ...; gp = &(o<~VPBlock(1,0)>.i.v) }
> ```
> The `~` prefix signals HeapUse/HeapDefine. Without the fix `o` is a plain stack-like `Define` and its NameExpr in `&o.i.v` has `NameExprTypeNormal`. With the fix, both are upgraded to `HeapDefine`/`HeapUse`.
> 
> **Why no observable crash:** the only code path that *requires* HeapItemValue is closure creation (it panics if a `HeapCaptures` entry is not a HeapItem). But `HeapCaptures` is populated by the closure analysis pass, which independently marks `o` for heap use whenever a closure captures it. When there's no closure, `o` never ends up in `HeapCaptures`, so no panic — and Go GC keeps the nested struct alive regardless.
> 
> **Suggested filetest:** The most useful thing the PR could add is a `// Preprocessed:` filetest documenting the expected annotations, rather than an output-only test. But this requires the unit test compile error to be fixed first.

I added a filetest, and fixed the CI. Please take a last look :) (fixing the current error)

thehowl committed:
fix(gnovm): rename slc helper param to avoid shadowing predeclared max
